[m0destly] iP

m0destly opened the pull request:
### Darwin
> "Thank you for your support." - Darwin Nunez

Darwin frees your mind of having to remember things you need to do. It's:

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is:
1. Download it from [here](https://github.com/m0destly/ip)
2. Double click it
3. Add your tasks
4. Let it manage your tasks for you 🤩 

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) 
##
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

m0destly committed:
Rename main class, add greeting

m0destly committed:
Add user input, echo commands

m0destly committed:
Added store text, list functions

m0destly committed:
Added mark, unmark functions and Task class

m0destly committed:
Added Todo, Deadline, Event classes

m0destly committed:
Added testing by I/O redirection

m0destly committed:
Added DarwinException (custom exception), TaskHandler classes to handle errors

m0destly committed:
Added delete function, minor bug fixes

m0destly committed:
Added ErrorMessage enum class, some bug fixes

m0destly committed:
Add saveTasks method, data directory containing save file

m0destly committed:
Add loadTasks method, print statements to indicate load status of tasks

m0destly committed:
Merge branch 'branch-level-7'

m0destly committed:
Changed deadline and event strings to LocalDate format

m0destly committed:
Change error messages to show correct format and usage

m0destly committed:
Merge branch 'branch-level-8'

m0destly committed:
Refactor code to use more OOP

m0destly committed:
Divide into packages

m0destly committed:
Merge branch 'add-gradle-support'

m0destly committed:
Add JUnit tests for parse method in TaskList class, load method in Storage class

m0destly committed:
Create JAR file

m0destly committed:
Add find function and minor bugfixes

m0destly committed:
Adjust code to adhere to coding standards

m0destly committed:
Resolve minor bug

m0destly committed:
Add JavaDocs for non-trivial main class methods

m0destly committed:
Merge branch 'branch-Level-9'

m0destly committed:
Merge branch 'branch-A-JavaDoc'

m0destly committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/darwin/Parser.java

m0destly committed:
Fix minor bug

m0destly committed:
Fix minor bug

Shamanbenny commented on src/main/java/darwin/Parser.java:
```diff
@@ -0,0 +1,69 @@
+                int taskNumber = Integer.parseInt(index) - 1;
+                taskList.mark(taskNumber);
+            } catch (NumberFormatException e) {
+                throw new DarwinException(ErrorMessage.NOT_NUMBER.message());
+            }
+        } else if (command.equals("unmark") || command.startsWith("unmark ")) {
+            try {
+                String index = command.substring(6).trim();
+                // No index
+                if (index.isEmpty()) {
+                    throw new DarwinException(ErrorMessage.MISSING_INDEX_UNMARK.message());
+                }
+                int taskNumber = Integer.parseInt(index) - 1;
+                taskList.unmark(taskNumber);
+            } catch (NumberFormatException e) {
+                throw new DarwinException(ErrorMessage.NOT_NUMBER.message());
+            }
+        } else if (command.equals("delete") || command.startsWith("delete ")) {
+            try {
+                String index = command.substring(6).trim();
+                // No index
+                if (index.isEmpty()) {
+                    throw new DarwinException(ErrorMessage.MISSING_INDEX_DELETE.message());
+                }
+                int taskNumber = Integer.parseInt(index) - 1;
+                taskList.delete(taskNumber);
+            } catch (NumberFormatException e) {
+                throw new DarwinException(ErrorMessage.NOT_NUMBER.message());
+            }
+        } else if (command.equals("find") || command.startsWith("find ")) {
+            String keyword = command.substring(4).trim();
+            if (keyword.isEmpty()) {
+                throw new DarwinException(ErrorMessage.MISSING_KEYWORD.message());
+            }
+            taskList.find(keyword);
+        } else {
+            String[] inputs = command.split(" /");
+            taskList.add(inputs);
+        }
+        return false;
```
I feel like it might be easier to grasp the entire structure of this if it was set in `switch case` instead?

Shamanbenny replied:
If I am also not mistaken, I believe `switch case` on `String` variable is done with Hash lookup, which results in an O(1) lookup rather than chaining `if else` statements which could result in O(n), where `n` is the number of conditions.
Read more about it [here](https://medium.com/@thedevelopernote/efficiency-of-if-vs-switch-in-java-a-quick-comparison-32f4ee0e20a3).

m0destly replied:
Thanks for the insightful suggestion 😄

Shamanbenny commented on src/main/java/darwin/Storage.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.ArrayList;
+
+public class Storage {
+    private String filePath;
```
Would it be better to use `final` to prevent accidental assignment?

Shamanbenny commented on src/main/java/darwin/Storage.java:
```diff
@@ -0,0 +1,78 @@
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.ArrayList;
```
Should these imports be ordered in accordance with the module's [Coding Standard](https://se-education.org/guides/tutorials/intellijCodeStyle.html#tweak-import-order)?

Shamanbenny commented on src/main/java/darwin/TaskList.java:
```diff
@@ -0,0 +1,220 @@
+     * @throws DarwinException If the format of the command is violated or command is not understood.
+     */
+    public void add(String[] inputs) throws DarwinException {
+        if (inputs[0].equals("todo") || inputs[0].startsWith("todo ")) {
```
Would it be better to extract these `if else` statements and instead, call them by their own functions instead? (E.g. `addTodo()`, etc.)

Shamanbenny commented on src/main/java/darwin/Ui.java:
```diff
@@ -0,0 +1,32 @@
+import java.util.Scanner;
+
+public class Ui {
+    private static final String divider = "____________________________________________________________";
```
Should this be in `SCREAMING_SNAKE_CASE` instead? Refer to [this](https://se-education.org/guides/conventions/java/intermediate.html#naming).

m0destly replied:
Good catch 👍

Shamanbenny commented on src/main/java/exception/ErrorMessage.java:
I like your intuitive idea to use enum for error messages!

Shamanbenny reviewed (commented):
Wonderful work thus far, aside from a few nits. Keep up the effort!

Fieash commented on src/main/java/darwin/Storage.java:
```diff
@@ -25,7 +25,7 @@ public ArrayList<Task> load() throws DarwinException {
             ArrayList<Task> tasks = (ArrayList<Task>) ois.readObject();
             ois.close();
-            System.out.println(tasks.size() + " tasks loaded.");
+            System.out.println(tasks.size() + (tasks.size() > 1 ? " tasks loaded." : " task loaded."));
```
I like the concise print statement, but should this be a one liner?

m0destly committed:
Adjust code to follow convention

m0destly committed:
Add checkstyle, suppressions files and enabled Checkstyle-IDEA plugin

m0destly committed:
Merge branch 'branch-A-CheckStyle'

Shamanbenny commented on src/main/java/darwin/Storage.java:
```diff
@@ -13,7 +13,7 @@
 
 public class Storage {
-    private String filePath;
+    private final String filePath;
```
Just glimpsed over your commits, as I was notified via email notification, I believe this should also be `FILE_PATH` as it's a constant variable.

Fieash reviewed (commented):
LGTM, some splitting of lines requested

m0destly committed:
Add GUI, change relevant methods

m0destly committed:
Merge branch 'branch-Level-10'

m0destly committed:
Add assert statements at handleUserInput() and load() methods

The handleUserInput() assert statement checks that the output is not empty. The load() assert statement checks that tasks being loaded do exist.

m0destly committed:
Adjust to adhere to coding standards

Replaced magic numbers with named constants, replaced complicated expressions with more readable code, and happy path is more prominent.

m0destly committed:
Merge pull request #1 from m0destly/A-Assertions

Add assert statements at handleUserInput() and load() methods

m0destly committed:
Merge branch 'master' of https://github.com/m0destly/ip

* 'master' of https://github.com/m0destly/ip:
  Add assert statements at handleUserInput() and load() methods

m0destly committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add assert statements at handleUserInput() and load() methods

# Conflicts:
#	src/main/java/darwin/Storage.java

m0destly committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Adjust to adhere to coding standards

m0destly committed:
Add help command to access list of all commands, fix tasks not saved on closing GUI window

m0destly committed:
Merge branch 'branch-C-Help'

* branch-C-Help:
  Add help command to access list of all commands, fix tasks not saved on closing GUI window

m0destly committed:
Add parseMark method to shorten parse method

m0destly committed:
Add parseUnmark method to shorten parse method

m0destly committed:
Add parseDelete method to shorten parse method

m0destly committed:
Add parseFind method to shorten parse method

m0destly committed:
Add parseAdd method to shorten parse method

m0destly committed:
Add individual addTask methods to replace add method

m0destly committed:
Fix bug that allowed event dates to violate chronological order

m0destly committed:
Add JavaDocs for new parse methods

m0destly committed:
Fix bye message not shown on exit, update README and add Ui.png

m0destly committed:
Update README and fix save tasks bug

m0destly committed:
Add JavaDocs, update saved file

m0destly committed:
Update gradle build for JAR file to work

m0destly committed:
Add citation for use of AI

m0destly committed:
Update JUnit tests

m0destly committed:
Update to save tasks after every successful command
